{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stage 1: Identify Continues Blocks of CpGs that are Correlated (R^2 >= 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(matrixStats)\n",
    "library(stringr)\n",
    "\n",
    "create_empty_table <- function(num_rows, num_cols) {\n",
    "  frame <- data.frame(matrix(NA, nrow = num_rows, ncol = num_cols))\n",
    "  return(frame)\n",
    "}\n",
    "as.numeric.factor <- function(x) {as.numeric(levels(x))[x]}\n",
    "\n",
    "\n",
    "min.corr <- 0.707107\n",
    "\n",
    "#Input\n",
    "chr_num = \"19\"# args[1]\n",
    "merged_coverage_files = paste(\"../input_files/gtex_raw_bins/chr\",chr_num,\"/ind\",sep=\"\")\n",
    "#tissue_coverage_files = paste(\"../input_files/gtex_raw_bins/chr\",chr_num,\"/tissue/\",sep=\"\")\n",
    "\n",
    "\n",
    "#file1 <- read.csv(\"./chr1/meth_coverage/bin_200bp/ind/13OW7.txt.gathered.bam.sorted.bismark.bins.csv\",header = T)\n",
    "#x = \"./input_files/gtex_raw_bins/chr5/ind/X261.txt.gathered.bam.sorted.bismark.cov.cyt.merged_CpG_evidence.bins.csv\"\n",
    "#x = \"../input_files/gtex_raw_bins/chr1/ind/chr1.X261.txt.gathered.bam.sorted.bismark.cov.cyt.merged_CpG_evidence.bins.csv\"\n",
    "#strsplit(strsplit(x,'\\\\.')[[1]][3],\"/\")[[1]][6]\n",
    "#strsplit(x,'\\\\.')[[1]][4]\n",
    "\n",
    "file_names <- dir(merged_coverage_files,full.names = T) \n",
    "df <- do.call(rbind,\n",
    "            lapply(file_names, \n",
    "                   function(x) cbind(read.csv(x),\n",
    "                                     #name=strsplit(strsplit(x,'\\\\.')[[1]][2],\"/\")[[1]][6]\n",
    "                                     name=strsplit(x,'\\\\.')[[1]][4]\n",
    "                   )\n",
    "            )\n",
    ")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>UCSC.Browser.Coordinates</th><th scope=col>Bin.Name</th><th scope=col>Avg.Methylation</th><th scope=col>CpGs.Covered</th><th scope=col>Num.of.CpGs.with.RD_n</th><th scope=col>name</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr19:68500-68600</td><td>chr19_68600</td><td>96.491228</td><td>3</td><td>3</td><td>13OW7</td></tr>\n",
       "\t<tr><td>chr19:70200-70300</td><td>chr19_70300</td><td> 1.123596</td><td>3</td><td>3</td><td>13OW7</td></tr>\n",
       "\t<tr><td>chr19:80600-80700</td><td>chr19_80700</td><td>85.000000</td><td>2</td><td>2</td><td>13OW7</td></tr>\n",
       "\t<tr><td>chr19:80900-81000</td><td>chr19_81000</td><td>92.063492</td><td>2</td><td>2</td><td>13OW7</td></tr>\n",
       "\t<tr><td>chr19:81600-81700</td><td>chr19_81700</td><td>62.857143</td><td>2</td><td>2</td><td>13OW7</td></tr>\n",
       "\t<tr><td>chr19:91000-91100</td><td>chr19_91100</td><td>69.285714</td><td>2</td><td>2</td><td>13OW7</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       " UCSC.Browser.Coordinates & Bin.Name & Avg.Methylation & CpGs.Covered & Num.of.CpGs.with.RD\\_n & name\\\\\n",
       " <fct> & <fct> & <dbl> & <int> & <int> & <fct>\\\\\n",
       "\\hline\n",
       "\t chr19:68500-68600 & chr19\\_68600 & 96.491228 & 3 & 3 & 13OW7\\\\\n",
       "\t chr19:70200-70300 & chr19\\_70300 &  1.123596 & 3 & 3 & 13OW7\\\\\n",
       "\t chr19:80600-80700 & chr19\\_80700 & 85.000000 & 2 & 2 & 13OW7\\\\\n",
       "\t chr19:80900-81000 & chr19\\_81000 & 92.063492 & 2 & 2 & 13OW7\\\\\n",
       "\t chr19:81600-81700 & chr19\\_81700 & 62.857143 & 2 & 2 & 13OW7\\\\\n",
       "\t chr19:91000-91100 & chr19\\_91100 & 69.285714 & 2 & 2 & 13OW7\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| UCSC.Browser.Coordinates &lt;fct&gt; | Bin.Name &lt;fct&gt; | Avg.Methylation &lt;dbl&gt; | CpGs.Covered &lt;int&gt; | Num.of.CpGs.with.RD_n &lt;int&gt; | name &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| chr19:68500-68600 | chr19_68600 | 96.491228 | 3 | 3 | 13OW7 |\n",
       "| chr19:70200-70300 | chr19_70300 |  1.123596 | 3 | 3 | 13OW7 |\n",
       "| chr19:80600-80700 | chr19_80700 | 85.000000 | 2 | 2 | 13OW7 |\n",
       "| chr19:80900-81000 | chr19_81000 | 92.063492 | 2 | 2 | 13OW7 |\n",
       "| chr19:81600-81700 | chr19_81700 | 62.857143 | 2 | 2 | 13OW7 |\n",
       "| chr19:91000-91100 | chr19_91100 | 69.285714 | 2 | 2 | 13OW7 |\n",
       "\n"
      ],
      "text/plain": [
       "  UCSC.Browser.Coordinates Bin.Name    Avg.Methylation CpGs.Covered\n",
       "1 chr19:68500-68600        chr19_68600 96.491228       3           \n",
       "2 chr19:70200-70300        chr19_70300  1.123596       3           \n",
       "3 chr19:80600-80700        chr19_80700 85.000000       2           \n",
       "4 chr19:80900-81000        chr19_81000 92.063492       2           \n",
       "5 chr19:81600-81700        chr19_81700 62.857143       2           \n",
       "6 chr19:91000-91100        chr19_91100 69.285714       2           \n",
       "  Num.of.CpGs.with.RD_n name \n",
       "1 3                     13OW7\n",
       "2 3                     13OW7\n",
       "3 2                     13OW7\n",
       "4 2                     13OW7\n",
       "5 2                     13OW7\n",
       "6 2                     13OW7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor(vec1, vec2, use = \"pairwise.complete.obs\", method = \"pearson\"):\n",
      "“the standard deviation is zero”Warning message in cor(vec1, vec2, use = \"pairwise.complete.obs\", method = \"pearson\"):\n",
      "“the standard deviation is zero”Warning message in cor(vec1, vec2, use = \"pairwise.complete.obs\", method = \"pearson\"):\n",
      "“the standard deviation is zero”Warning message in cor(vec1, vec2, use = \"pairwise.complete.obs\", method = \"pearson\"):\n",
      "“the standard deviation is zero”Warning message in cor(vec1, vec2, use = \"pairwise.complete.obs\", method = \"pearson\"):\n",
      "“the standard deviation is zero”"
     ]
    }
   ],
   "source": [
    "individuals <- as.character(unique(df$name))\n",
    "I <- split(df, df$name)\n",
    "combined_I <- I %>% Reduce(function(dtf1,dtf2) full_join(dtf1,dtf2,by=\"Bin.Name\"), .)\n",
    "combined_I <- na.omit(combined_I)\n",
    "avg.m<- combined_I[ , grepl( \"Avg.Methy\" , names( combined_I ))]\n",
    "combined_I$IIV <- rowVars(as.matrix(avg.m))\n",
    "temp_min_max <- as.data.frame(rowRanges(as.matrix(avg.m),na.rm = TRUE ))\n",
    "combined_I$IIR <- abs(temp_min_max$V1-temp_min_max$V2)\n",
    "sortedcombined_I <- combined_I[order(-combined_I$IIR),]#sorted by the range\n",
    "sortedcombined_I <-na.omit(sortedcombined_I)\n",
    "avg.methy_df_ind <- sortedcombined_I[,c(grep(\"Bin.|Avg.\",names(sortedcombined_I))), drop=F]\n",
    "colnames(avg.methy_df_ind)[2:11] <- individuals\n",
    "\n",
    "#combined_I$IIV <- row(combined_I[ , grepl( \"Avg.Methy\" , names( combined_I ) ) ], na.rm=TRUE)\n",
    "\n",
    "#write.csv(x = sortedcombined_I,file = \"temp_stage1.csv\",row.names = F)\n",
    "#extract top %20 of the bins\n",
    "avg.methy_df_ind <- sortedcombined_I[,c(grep(\"Bin.|Avg.\",names(sortedcombined_I))), drop=F]\n",
    "\n",
    "colnames(avg.methy_df_ind)[2:11] <- individuals\n",
    "\n",
    "\n",
    "split_Bin.name <- data.frame(str_split_fixed(avg.methy_df_ind$Bin.Name, \"_\", 2))\n",
    "avg.methy_df_ind <- cbind(avg.methy_df_ind,split_Bin.name)\n",
    "avg.methy_df_ind$X1 <- NULL\n",
    "avg.methy_df_ind$X2 <- as.numeric(levels(avg.methy_df_ind$X2))[avg.methy_df_ind$X2]\n",
    "avg.methy_df_ind <- avg.methy_df_ind[order(avg.methy_df_ind$X2),]\n",
    "\n",
    "#avg.methy_df_ind <- avg.methy_df_ind[1:100,]\n",
    "\n",
    "number_of_bins <- dim(avg.methy_df_ind)[1]\n",
    "rownames(avg.methy_df_ind) <- NULL\n",
    "\n",
    "\n",
    "i=1\n",
    "min.dist <- 200\n",
    "collect_index <- c()\n",
    "cluster_num <- 0\n",
    "cluster_vec <- c()\n",
    "                           \n",
    "while(i < number_of_bins){\n",
    "    #print(i)\n",
    "    rec_i <- avg.methy_df_ind[i,]\n",
    "    rec_j <- avg.methy_df_ind[i+1,]\n",
    "\n",
    "    if(as.numeric(rec_j[,12])-as.numeric(rec_i[,12]) > min.dist){\n",
    "        i <- i + 1\n",
    "        next\n",
    "    }\n",
    "\n",
    "    vec1 <- as.numeric(rec_i[,c(2:11)])\n",
    "    vec2 <- as.numeric(rec_j[,c(2:11)])\n",
    "    pair_corr <- try(\n",
    "    expr = cor(vec1,vec2,use = \"pairwise.complete.obs\",method =\"pearson\")\n",
    "    )\n",
    "    if(is.na(pair_corr)){pair_corr <- 0}\n",
    "\n",
    "    if(pair_corr < min.corr){i <- i + 1 \n",
    "                             next }\n",
    "\n",
    "\n",
    "    if(i %in% collect_index ){\n",
    "        cluster_vec <- c(cluster_vec,rep(cluster_num,2))\n",
    "    }else{\n",
    "        cluster_num <- cluster_num + 1\n",
    "        cluster_vec <- c(cluster_vec,rep(cluster_num,2))\n",
    "    }\n",
    "    collect_index <- c(collect_index,c(i,i+1))\n",
    "    i <- i + 1\n",
    "}\n",
    "\n",
    "\n",
    "avg.methy_df_ind_clustered <- avg.methy_df_ind[collect_index,]\n",
    "avg.methy_df_ind_clustered$cluser_ID <- cluster_vec\n",
    "avg.methy_df_ind_clustered <- avg.methy_df_ind_clustered[!duplicated(avg.methy_df_ind_clustered$Bin.Name),]\n",
    "\n",
    "write.csv(avg.methy_df_ind_clustered,file = paste(\"../predefined_R2_0.5/chr\",chr_num,\"_\",min.corr,\".csv\",sep=\"\"),row.names = F)\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 20 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Bin.Name</th><th scope=col>13OW7</th><th scope=col>13QIC</th><th scope=col>14PQA</th><th scope=col>1F52S</th><th scope=col>1F6RS</th><th scope=col>1GZHY</th><th scope=col>RVPV</th><th scope=col>S7SE</th><th scope=col>WVLH</th><th scope=col>X261</th><th scope=col>X2</th><th scope=col>cluser_ID</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>17</th><td>chr19_93300 </td><td>97.663551</td><td>87.434555</td><td>83.934426</td><td>91.883117</td><td>92.1296296</td><td>93.055556</td><td>90.714286</td><td>93.225806</td><td>92.727273</td><td>89.944134</td><td> 93300</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>chr19_93400 </td><td>95.753425</td><td>89.236111</td><td>85.787848</td><td>91.658391</td><td>88.8888889</td><td>94.456763</td><td>88.396349</td><td>94.905386</td><td>92.625899</td><td>87.357631</td><td> 93400</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>chr19_93500 </td><td>93.958076</td><td>87.735849</td><td>86.235187</td><td>91.107078</td><td>90.6647808</td><td>93.159609</td><td>88.061336</td><td>92.307692</td><td>89.825971</td><td>87.860697</td><td> 93500</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>chr19_93600 </td><td>84.641256</td><td>74.688797</td><td>79.239130</td><td>81.589537</td><td>82.3899371</td><td>80.529595</td><td>81.733021</td><td>87.648456</td><td>83.445491</td><td>74.206755</td><td> 93600</td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>chr19_93700 </td><td>95.087719</td><td>92.264151</td><td>93.189557</td><td>93.776371</td><td>95.0067476</td><td>95.432300</td><td>94.276795</td><td>96.828992</td><td>92.571429</td><td>92.045455</td><td> 93700</td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>85</th><td>chr19_200400</td><td>95.864662</td><td>96.047431</td><td>92.401216</td><td>93.763920</td><td>95.2112676</td><td>93.582888</td><td>89.825581</td><td>94.301994</td><td>97.487437</td><td>94.662921</td><td>200400</td><td> 3</td></tr>\n",
       "\t<tr><th scope=row>86</th><td>chr19_200500</td><td>87.000000</td><td>84.615385</td><td>74.887892</td><td>73.228346</td><td>77.4774775</td><td>87.500000</td><td>68.103448</td><td>80.314961</td><td>89.855072</td><td>77.894737</td><td>200500</td><td> 3</td></tr>\n",
       "\t<tr><th scope=row>126</th><td>chr19_240700</td><td>84.023669</td><td>81.280788</td><td>84.444444</td><td>82.800000</td><td>73.3766234</td><td>86.187845</td><td>84.516129</td><td>89.603960</td><td>80.158730</td><td>85.500000</td><td>240700</td><td> 4</td></tr>\n",
       "\t<tr><th scope=row>127</th><td>chr19_240800</td><td>87.037037</td><td>91.208791</td><td>94.160584</td><td>88.957055</td><td>85.6164384</td><td>93.491124</td><td>90.000000</td><td>93.750000</td><td>87.387387</td><td>91.860465</td><td>240800</td><td> 4</td></tr>\n",
       "\t<tr><th scope=row>144</th><td>chr19_245900</td><td>55.932203</td><td>77.173913</td><td>64.516129</td><td>77.142857</td><td>77.2151899</td><td>93.333333</td><td>66.129032</td><td>55.263158</td><td>67.213115</td><td>69.014085</td><td>245900</td><td> 5</td></tr>\n",
       "\t<tr><th scope=row>145</th><td>chr19_246100</td><td>34.117647</td><td>34.117647</td><td>24.444444</td><td>34.000000</td><td>32.5842697</td><td>43.181818</td><td>26.582278</td><td>22.580645</td><td>32.835821</td><td>37.179487</td><td>246100</td><td> 5</td></tr>\n",
       "\t<tr><th scope=row>150</th><td>chr19_246600</td><td>75.753769</td><td>76.470588</td><td>72.506990</td><td>83.071342</td><td>77.4553571</td><td>75.636943</td><td>74.766355</td><td>80.649351</td><td>76.629571</td><td>78.940568</td><td>246600</td><td> 6</td></tr>\n",
       "\t<tr><th scope=row>151</th><td>chr19_246700</td><td>70.154374</td><td>69.031142</td><td>62.790698</td><td>73.170732</td><td>69.9398798</td><td>66.257669</td><td>63.789474</td><td>69.142857</td><td>67.224880</td><td>65.969582</td><td>246700</td><td> 6</td></tr>\n",
       "\t<tr><th scope=row>154</th><td>chr19_247000</td><td>44.186047</td><td>42.123288</td><td>36.578947</td><td>40.000000</td><td>43.6746988</td><td>42.729970</td><td>31.487889</td><td>37.549407</td><td>48.771930</td><td>43.150685</td><td>247000</td><td> 7</td></tr>\n",
       "\t<tr><th scope=row>155</th><td>chr19_247100</td><td>56.431535</td><td>54.578755</td><td>46.749226</td><td>53.846154</td><td>55.5970149</td><td>48.000000</td><td>45.614035</td><td>53.333333</td><td>57.089552</td><td>59.322034</td><td>247100</td><td> 7</td></tr>\n",
       "\t<tr><th scope=row>254</th><td>chr19_267100</td><td> 3.125000</td><td> 3.649635</td><td> 6.206897</td><td> 1.204819</td><td> 0.7575758</td><td> 1.360544</td><td> 4.166667</td><td> 1.652893</td><td> 2.542373</td><td> 1.418440</td><td>267100</td><td> 8</td></tr>\n",
       "\t<tr><th scope=row>255</th><td>chr19_267200</td><td> 3.064067</td><td> 2.605210</td><td> 5.515088</td><td> 1.464255</td><td> 2.5462963</td><td> 1.368421</td><td> 4.031008</td><td> 1.750292</td><td> 1.241379</td><td> 1.055662</td><td>267200</td><td> 8</td></tr>\n",
       "\t<tr><th scope=row>261</th><td>chr19_267800</td><td>28.066528</td><td>35.368957</td><td>27.407407</td><td>43.462898</td><td>30.2721088</td><td>30.530165</td><td>29.257642</td><td>30.597015</td><td>25.296443</td><td>29.173989</td><td>267800</td><td> 9</td></tr>\n",
       "\t<tr><th scope=row>262</th><td>chr19_267900</td><td>56.678700</td><td>66.666667</td><td>56.826568</td><td>71.729958</td><td>69.0839695</td><td>61.200000</td><td>58.426966</td><td>62.301587</td><td>59.375000</td><td>62.264151</td><td>267900</td><td> 9</td></tr>\n",
       "\t<tr><th scope=row>290</th><td>chr19_272100</td><td>63.000000</td><td>63.500000</td><td>66.666667</td><td>63.594470</td><td>66.6666667</td><td>67.241379</td><td>64.814815</td><td>55.208333</td><td>67.179487</td><td>70.305677</td><td>272100</td><td>10</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 20 × 13\n",
       "\\begin{tabular}{r|lllllllllllll}\n",
       "  & Bin.Name & 13OW7 & 13QIC & 14PQA & 1F52S & 1F6RS & 1GZHY & RVPV & S7SE & WVLH & X261 & X2 & cluser\\_ID\\\\\n",
       "  & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t17 & chr19\\_93300  & 97.663551 & 87.434555 & 83.934426 & 91.883117 & 92.1296296 & 93.055556 & 90.714286 & 93.225806 & 92.727273 & 89.944134 &  93300 &  1\\\\\n",
       "\t18 & chr19\\_93400  & 95.753425 & 89.236111 & 85.787848 & 91.658391 & 88.8888889 & 94.456763 & 88.396349 & 94.905386 & 92.625899 & 87.357631 &  93400 &  1\\\\\n",
       "\t19 & chr19\\_93500  & 93.958076 & 87.735849 & 86.235187 & 91.107078 & 90.6647808 & 93.159609 & 88.061336 & 92.307692 & 89.825971 & 87.860697 &  93500 &  1\\\\\n",
       "\t20 & chr19\\_93600  & 84.641256 & 74.688797 & 79.239130 & 81.589537 & 82.3899371 & 80.529595 & 81.733021 & 87.648456 & 83.445491 & 74.206755 &  93600 &  2\\\\\n",
       "\t21 & chr19\\_93700  & 95.087719 & 92.264151 & 93.189557 & 93.776371 & 95.0067476 & 95.432300 & 94.276795 & 96.828992 & 92.571429 & 92.045455 &  93700 &  2\\\\\n",
       "\t85 & chr19\\_200400 & 95.864662 & 96.047431 & 92.401216 & 93.763920 & 95.2112676 & 93.582888 & 89.825581 & 94.301994 & 97.487437 & 94.662921 & 200400 &  3\\\\\n",
       "\t86 & chr19\\_200500 & 87.000000 & 84.615385 & 74.887892 & 73.228346 & 77.4774775 & 87.500000 & 68.103448 & 80.314961 & 89.855072 & 77.894737 & 200500 &  3\\\\\n",
       "\t126 & chr19\\_240700 & 84.023669 & 81.280788 & 84.444444 & 82.800000 & 73.3766234 & 86.187845 & 84.516129 & 89.603960 & 80.158730 & 85.500000 & 240700 &  4\\\\\n",
       "\t127 & chr19\\_240800 & 87.037037 & 91.208791 & 94.160584 & 88.957055 & 85.6164384 & 93.491124 & 90.000000 & 93.750000 & 87.387387 & 91.860465 & 240800 &  4\\\\\n",
       "\t144 & chr19\\_245900 & 55.932203 & 77.173913 & 64.516129 & 77.142857 & 77.2151899 & 93.333333 & 66.129032 & 55.263158 & 67.213115 & 69.014085 & 245900 &  5\\\\\n",
       "\t145 & chr19\\_246100 & 34.117647 & 34.117647 & 24.444444 & 34.000000 & 32.5842697 & 43.181818 & 26.582278 & 22.580645 & 32.835821 & 37.179487 & 246100 &  5\\\\\n",
       "\t150 & chr19\\_246600 & 75.753769 & 76.470588 & 72.506990 & 83.071342 & 77.4553571 & 75.636943 & 74.766355 & 80.649351 & 76.629571 & 78.940568 & 246600 &  6\\\\\n",
       "\t151 & chr19\\_246700 & 70.154374 & 69.031142 & 62.790698 & 73.170732 & 69.9398798 & 66.257669 & 63.789474 & 69.142857 & 67.224880 & 65.969582 & 246700 &  6\\\\\n",
       "\t154 & chr19\\_247000 & 44.186047 & 42.123288 & 36.578947 & 40.000000 & 43.6746988 & 42.729970 & 31.487889 & 37.549407 & 48.771930 & 43.150685 & 247000 &  7\\\\\n",
       "\t155 & chr19\\_247100 & 56.431535 & 54.578755 & 46.749226 & 53.846154 & 55.5970149 & 48.000000 & 45.614035 & 53.333333 & 57.089552 & 59.322034 & 247100 &  7\\\\\n",
       "\t254 & chr19\\_267100 &  3.125000 &  3.649635 &  6.206897 &  1.204819 &  0.7575758 &  1.360544 &  4.166667 &  1.652893 &  2.542373 &  1.418440 & 267100 &  8\\\\\n",
       "\t255 & chr19\\_267200 &  3.064067 &  2.605210 &  5.515088 &  1.464255 &  2.5462963 &  1.368421 &  4.031008 &  1.750292 &  1.241379 &  1.055662 & 267200 &  8\\\\\n",
       "\t261 & chr19\\_267800 & 28.066528 & 35.368957 & 27.407407 & 43.462898 & 30.2721088 & 30.530165 & 29.257642 & 30.597015 & 25.296443 & 29.173989 & 267800 &  9\\\\\n",
       "\t262 & chr19\\_267900 & 56.678700 & 66.666667 & 56.826568 & 71.729958 & 69.0839695 & 61.200000 & 58.426966 & 62.301587 & 59.375000 & 62.264151 & 267900 &  9\\\\\n",
       "\t290 & chr19\\_272100 & 63.000000 & 63.500000 & 66.666667 & 63.594470 & 66.6666667 & 67.241379 & 64.814815 & 55.208333 & 67.179487 & 70.305677 & 272100 & 10\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 20 × 13\n",
       "\n",
       "| <!--/--> | Bin.Name &lt;fct&gt; | 13OW7 &lt;dbl&gt; | 13QIC &lt;dbl&gt; | 14PQA &lt;dbl&gt; | 1F52S &lt;dbl&gt; | 1F6RS &lt;dbl&gt; | 1GZHY &lt;dbl&gt; | RVPV &lt;dbl&gt; | S7SE &lt;dbl&gt; | WVLH &lt;dbl&gt; | X261 &lt;dbl&gt; | X2 &lt;dbl&gt; | cluser_ID &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 17 | chr19_93300  | 97.663551 | 87.434555 | 83.934426 | 91.883117 | 92.1296296 | 93.055556 | 90.714286 | 93.225806 | 92.727273 | 89.944134 |  93300 |  1 |\n",
       "| 18 | chr19_93400  | 95.753425 | 89.236111 | 85.787848 | 91.658391 | 88.8888889 | 94.456763 | 88.396349 | 94.905386 | 92.625899 | 87.357631 |  93400 |  1 |\n",
       "| 19 | chr19_93500  | 93.958076 | 87.735849 | 86.235187 | 91.107078 | 90.6647808 | 93.159609 | 88.061336 | 92.307692 | 89.825971 | 87.860697 |  93500 |  1 |\n",
       "| 20 | chr19_93600  | 84.641256 | 74.688797 | 79.239130 | 81.589537 | 82.3899371 | 80.529595 | 81.733021 | 87.648456 | 83.445491 | 74.206755 |  93600 |  2 |\n",
       "| 21 | chr19_93700  | 95.087719 | 92.264151 | 93.189557 | 93.776371 | 95.0067476 | 95.432300 | 94.276795 | 96.828992 | 92.571429 | 92.045455 |  93700 |  2 |\n",
       "| 85 | chr19_200400 | 95.864662 | 96.047431 | 92.401216 | 93.763920 | 95.2112676 | 93.582888 | 89.825581 | 94.301994 | 97.487437 | 94.662921 | 200400 |  3 |\n",
       "| 86 | chr19_200500 | 87.000000 | 84.615385 | 74.887892 | 73.228346 | 77.4774775 | 87.500000 | 68.103448 | 80.314961 | 89.855072 | 77.894737 | 200500 |  3 |\n",
       "| 126 | chr19_240700 | 84.023669 | 81.280788 | 84.444444 | 82.800000 | 73.3766234 | 86.187845 | 84.516129 | 89.603960 | 80.158730 | 85.500000 | 240700 |  4 |\n",
       "| 127 | chr19_240800 | 87.037037 | 91.208791 | 94.160584 | 88.957055 | 85.6164384 | 93.491124 | 90.000000 | 93.750000 | 87.387387 | 91.860465 | 240800 |  4 |\n",
       "| 144 | chr19_245900 | 55.932203 | 77.173913 | 64.516129 | 77.142857 | 77.2151899 | 93.333333 | 66.129032 | 55.263158 | 67.213115 | 69.014085 | 245900 |  5 |\n",
       "| 145 | chr19_246100 | 34.117647 | 34.117647 | 24.444444 | 34.000000 | 32.5842697 | 43.181818 | 26.582278 | 22.580645 | 32.835821 | 37.179487 | 246100 |  5 |\n",
       "| 150 | chr19_246600 | 75.753769 | 76.470588 | 72.506990 | 83.071342 | 77.4553571 | 75.636943 | 74.766355 | 80.649351 | 76.629571 | 78.940568 | 246600 |  6 |\n",
       "| 151 | chr19_246700 | 70.154374 | 69.031142 | 62.790698 | 73.170732 | 69.9398798 | 66.257669 | 63.789474 | 69.142857 | 67.224880 | 65.969582 | 246700 |  6 |\n",
       "| 154 | chr19_247000 | 44.186047 | 42.123288 | 36.578947 | 40.000000 | 43.6746988 | 42.729970 | 31.487889 | 37.549407 | 48.771930 | 43.150685 | 247000 |  7 |\n",
       "| 155 | chr19_247100 | 56.431535 | 54.578755 | 46.749226 | 53.846154 | 55.5970149 | 48.000000 | 45.614035 | 53.333333 | 57.089552 | 59.322034 | 247100 |  7 |\n",
       "| 254 | chr19_267100 |  3.125000 |  3.649635 |  6.206897 |  1.204819 |  0.7575758 |  1.360544 |  4.166667 |  1.652893 |  2.542373 |  1.418440 | 267100 |  8 |\n",
       "| 255 | chr19_267200 |  3.064067 |  2.605210 |  5.515088 |  1.464255 |  2.5462963 |  1.368421 |  4.031008 |  1.750292 |  1.241379 |  1.055662 | 267200 |  8 |\n",
       "| 261 | chr19_267800 | 28.066528 | 35.368957 | 27.407407 | 43.462898 | 30.2721088 | 30.530165 | 29.257642 | 30.597015 | 25.296443 | 29.173989 | 267800 |  9 |\n",
       "| 262 | chr19_267900 | 56.678700 | 66.666667 | 56.826568 | 71.729958 | 69.0839695 | 61.200000 | 58.426966 | 62.301587 | 59.375000 | 62.264151 | 267900 |  9 |\n",
       "| 290 | chr19_272100 | 63.000000 | 63.500000 | 66.666667 | 63.594470 | 66.6666667 | 67.241379 | 64.814815 | 55.208333 | 67.179487 | 70.305677 | 272100 | 10 |\n",
       "\n"
      ],
      "text/plain": [
       "    Bin.Name     13OW7     13QIC     14PQA     1F52S     1F6RS      1GZHY    \n",
       "17  chr19_93300  97.663551 87.434555 83.934426 91.883117 92.1296296 93.055556\n",
       "18  chr19_93400  95.753425 89.236111 85.787848 91.658391 88.8888889 94.456763\n",
       "19  chr19_93500  93.958076 87.735849 86.235187 91.107078 90.6647808 93.159609\n",
       "20  chr19_93600  84.641256 74.688797 79.239130 81.589537 82.3899371 80.529595\n",
       "21  chr19_93700  95.087719 92.264151 93.189557 93.776371 95.0067476 95.432300\n",
       "85  chr19_200400 95.864662 96.047431 92.401216 93.763920 95.2112676 93.582888\n",
       "86  chr19_200500 87.000000 84.615385 74.887892 73.228346 77.4774775 87.500000\n",
       "126 chr19_240700 84.023669 81.280788 84.444444 82.800000 73.3766234 86.187845\n",
       "127 chr19_240800 87.037037 91.208791 94.160584 88.957055 85.6164384 93.491124\n",
       "144 chr19_245900 55.932203 77.173913 64.516129 77.142857 77.2151899 93.333333\n",
       "145 chr19_246100 34.117647 34.117647 24.444444 34.000000 32.5842697 43.181818\n",
       "150 chr19_246600 75.753769 76.470588 72.506990 83.071342 77.4553571 75.636943\n",
       "151 chr19_246700 70.154374 69.031142 62.790698 73.170732 69.9398798 66.257669\n",
       "154 chr19_247000 44.186047 42.123288 36.578947 40.000000 43.6746988 42.729970\n",
       "155 chr19_247100 56.431535 54.578755 46.749226 53.846154 55.5970149 48.000000\n",
       "254 chr19_267100  3.125000  3.649635  6.206897  1.204819  0.7575758  1.360544\n",
       "255 chr19_267200  3.064067  2.605210  5.515088  1.464255  2.5462963  1.368421\n",
       "261 chr19_267800 28.066528 35.368957 27.407407 43.462898 30.2721088 30.530165\n",
       "262 chr19_267900 56.678700 66.666667 56.826568 71.729958 69.0839695 61.200000\n",
       "290 chr19_272100 63.000000 63.500000 66.666667 63.594470 66.6666667 67.241379\n",
       "    RVPV      S7SE      WVLH      X261      X2     cluser_ID\n",
       "17  90.714286 93.225806 92.727273 89.944134  93300  1       \n",
       "18  88.396349 94.905386 92.625899 87.357631  93400  1       \n",
       "19  88.061336 92.307692 89.825971 87.860697  93500  1       \n",
       "20  81.733021 87.648456 83.445491 74.206755  93600  2       \n",
       "21  94.276795 96.828992 92.571429 92.045455  93700  2       \n",
       "85  89.825581 94.301994 97.487437 94.662921 200400  3       \n",
       "86  68.103448 80.314961 89.855072 77.894737 200500  3       \n",
       "126 84.516129 89.603960 80.158730 85.500000 240700  4       \n",
       "127 90.000000 93.750000 87.387387 91.860465 240800  4       \n",
       "144 66.129032 55.263158 67.213115 69.014085 245900  5       \n",
       "145 26.582278 22.580645 32.835821 37.179487 246100  5       \n",
       "150 74.766355 80.649351 76.629571 78.940568 246600  6       \n",
       "151 63.789474 69.142857 67.224880 65.969582 246700  6       \n",
       "154 31.487889 37.549407 48.771930 43.150685 247000  7       \n",
       "155 45.614035 53.333333 57.089552 59.322034 247100  7       \n",
       "254  4.166667  1.652893  2.542373  1.418440 267100  8       \n",
       "255  4.031008  1.750292  1.241379  1.055662 267200  8       \n",
       "261 29.257642 30.597015 25.296443 29.173989 267800  9       \n",
       "262 58.426966 62.301587 59.375000 62.264151 267900  9       \n",
       "290 64.814815 55.208333 67.179487 70.305677 272100 10       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(avg.methy_df_ind_clustered,20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stage 2: Calculate Inter-Tissue_Correlations in identified Blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# import libraries and load all data\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from scipy import stats\n",
    "import glob\n",
    "from itertools import chain\n",
    "\n",
    "chromosome=\"chr19\" #sys.argv[1]\n",
    "\n",
    "path = '../input_files/gtex_raw_covs/'+chromosome+'/*.cov'\n",
    "print(path)\n",
    "\n",
    "def get_data(path):\n",
    "    cov_files = glob.glob(path)\n",
    "    df = pd.DataFrame()\n",
    "    for file_ in cov_files:\n",
    "        file_df = pd.read_csv(file_,sep='\\t',header=None )\n",
    "        file_df['file_name'] = \"-\".join(os.path.basename(file_).split(\".\")[0].split(\"-\")[0:3])\n",
    "        df = df.append(file_df)\n",
    "\n",
    "    sample_name_anno = pd.read_csv(\"../annotations/sample_name_annotation.csv\",header = None)\n",
    "    sample_name_anno.columns = [\"sample_name\",\"tissue_type\"]\n",
    "    df.columns = [\"chr\",\"start\",\"end\",\"perc\",\"nMeth\",\"nUnmeth\",\"sample_name\"]\n",
    "    individuals = list(set([sample_name_anno.sample_name[i].split(\"-\")[1] for i in range(0,30)]))\n",
    "\n",
    "    all_sample_df = pd.merge(df,sample_name_anno,on = \"sample_name\")\n",
    "\n",
    "    #get methylation values for the block in 10 thyroid samples\n",
    "    Heart = all_sample_df[all_sample_df['tissue_type']==\"Heart\"]\n",
    "    Brain = all_sample_df[all_sample_df['tissue_type']==\"Brain\"]\n",
    "    Thyroid = all_sample_df[all_sample_df['tissue_type']==\"Thyroid\"]\n",
    "    return([Heart,Brain,Thyroid,individuals])\n",
    "\n",
    "\n",
    "def get_corr(start,end):\n",
    "    heart_values = []\n",
    "    brain_values = []\n",
    "    thyroid_values = []\n",
    "    sub_h = Heart[(start <= Heart['start']) & (Heart['end'] <= end)]\n",
    "    sub_b = Brain[(start <= Brain['start']) & (Brain['end'] <= end)]\n",
    "    sub_t = Thyroid[(start <= Thyroid['start']) & (Thyroid['end'] <= end)]\n",
    "\n",
    "\n",
    "    for ind in individuals:\n",
    "        temp1 = sub_h[sub_h['sample_name'].str.contains(ind)]\n",
    "        heart_values.append(get_meth(temp1))\n",
    "\n",
    "        temp2 = sub_b[sub_b['sample_name'].str.contains(ind)]\n",
    "        brain_values.append(get_meth(temp2))\n",
    "\n",
    "        temp3 = sub_t[sub_t['sample_name'].str.contains(ind)]\n",
    "        thyroid_values.append(get_meth(temp3))\n",
    "\n",
    "\n",
    "    block_methylation_df = pd.DataFrame(\n",
    "    {'heart': heart_values,\n",
    "     'brain': brain_values,\n",
    "     'thyroid': thyroid_values\n",
    "    })\n",
    "    block_methylation_df = block_methylation_df[['heart','brain','thyroid']]\n",
    "    cor_pairs = block_methylation_df.corr(method='pearson', min_periods=7)\n",
    "    return([block_methylation_df,cor_pairs])\n",
    "\n",
    "def get_meth(block_sample):\n",
    "    if(np.ceil(block_sample.shape[0]/2) < block_sample[(block_sample['nMeth'] + block_sample['nUnmeth']) >=5].shape[0]):\n",
    "        return(block_sample['perc'].mean())\n",
    "    else:\n",
    "        return (np.nan)\n",
    "\n",
    "\n",
    "[Heart,Brain,Thyroid,individuals] = get_data(path)\n",
    "for i in [0.707107]:\n",
    "    corr = str(i)\n",
    "    blocks = pd.read_csv(\"../predefined_R2_0.5/\"+chromosome+\"_\"+corr+\".csv\")\n",
    "    blocks['cluser_ID'] = blocks['cluser_ID'].astype(np.int64)\n",
    "    h_b_list = []\n",
    "    h_t_list = []\n",
    "    b_t_list = []\n",
    "    for i in range(1,blocks['cluser_ID'].iloc[-1]):\n",
    "        #print(i)\n",
    "        temp_range = blocks[blocks['cluser_ID']==i].X2\n",
    "        start = temp_range.iloc[0]-100\n",
    "        end = temp_range.iloc[-1]\n",
    "        temp_corr_mat=get_corr(start,end)[1]\n",
    "        h_b_list.append(temp_corr_mat.iloc[0,1])\n",
    "        h_t_list.append(temp_corr_mat.iloc[0,2])\n",
    "        b_t_list.append(temp_corr_mat.iloc[1,2])  \n",
    "\n",
    "    ITC_DF = pd.DataFrame.from_dict({'block_id':list(range(1,i+1)),\n",
    "                                 'Heart_Brain':h_b_list,\n",
    "                                 'Heart_Thyroid':h_t_list,\n",
    "                                 'Brain_Thyroid':b_t_list},\n",
    "                               )\n",
    "\n",
    "    ITC_DF = ITC_DF[['block_id','Heart_Brain','Heart_Thyroid','Brain_Thyroid']]\n",
    "\n",
    "    ITC_DF.to_csv(\"../predefined_R2_0.5/\"+chromosome+\"_\"+corr+\"_ITC.csv\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
